File: samples/java/TicTacToe/Client/src/main/java/systems/zlink/samples/tictactoe/client/TicTacToeClientArguments.java
package systems.zlink.samples.tictactoe.client;

public final class TicTacToeClientArguments {
    private TicTacToeClientArguments() {}

    public static TicTacToeClientOptions parse(String[] args) {
        TicTacToeClientOptions defaults = TicTacToeClientOptions.createDefault();
        return new TicTacToeClientOptions(
                readOption(args, "--api-url", defaults.apiUrl()),
                readOption(args, "--game-name", defaults.gameName()),
                readOption(args, "--x-actor-id", defaults.xActorId()),
                readOption(args, "--o-actor-id", defaults.oActorId()),
                readOption(args, "--observer-actor-id", defaults.observerActorId()),
                readOption(
                        args, "--lifecycle-completion-file", defaults.lifecycleCompletionFile()));
    }

    private static String readOption(String[] args, String name, String defaultValue) {
        for (int index = 0; index < args.length; index++) {
            if (!args[index].equals(name)) {
                continue;
            }
            if (index + 1 >= args.length) {
                throw new IllegalArgumentException("Missing value for '" + name + "'.");
            }
            return args[index + 1];
        }
        return defaultValue;
    }
}

File: samples/kotlin/TicTacToe/Client/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/client/TicTacToeClientArguments.kt
package systems.zlink.samples.kotlin.tictactoe.client

object TicTacToeClientArguments {
    fun parse(args: Array<String>): TicTacToeClientOptions {
        val defaults = TicTacToeClientOptions.createDefault()
        return TicTacToeClientOptions(
            apiUrl = readOption(args, "--api-url") ?: defaults.apiUrl,
            gameName = readOption(args, "--game-name") ?: defaults.gameName,
            xActorId = readOption(args, "--x-actor-id") ?: defaults.xActorId,
            oActorId = readOption(args, "--o-actor-id") ?: defaults.oActorId,
            observerActorId = readOption(args, "--observer-actor-id") ?: defaults.observerActorId,
            lifecycleCompletionFile =
                readOption(args, "--lifecycle-completion-file") ?: defaults.lifecycleCompletionFile,
        )
    }

    private fun readOption(args: Array<String>, name: String): String? {
        val index = args.indexOf(name)
        if (index < 0) {
            return null
        }
        require(index + 1 < args.size) { "Missing value for '$name'." }
        return args[index + 1]
    }
}
